sql/expression/function/soundex.go

```diff
@@ -0,0 +1,102 @@
+package function
+
+import (
+	"fmt"
+	"strings"
+	"unicode"
+
+	"gopkg.in/src-d/go-mysql-server.v0/sql"
+	"gopkg.in/src-d/go-mysql-server.v0/sql/expression"
+)
+
+// Soundex is a function that returns the soundex of a string. Two strings that
+// sound almost the same should have identical soundex strings. A standard
+// soundex string is four characters long, but the SOUNDEX() function returns
+// an arbitrarily long string.
+type Soundex struct {
+	expression.UnaryExpression
+}
+
+// NewSoundex creates a new Soundex expression.
+func NewSoundex(e sql.Expression) sql.Expression {
+	return &Soundex{expression.UnaryExpression{Child: e}}
+}
+
+// Eval implements the Expression interface.
+func (s *Soundex) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
+	v, err := s.Child.Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+
+	if v == nil {
+		return nil, nil
+	}
+
+	v, err = sql.Text.Convert(v)
+	if err != nil {
+		return nil, err
+	}
+
+	var b strings.Builder
+	var last rune
+	for _, c := range strings.ToUpper(v.(string)) {
+		if last == 0 && !unicode.IsLetter(c) {
+			continue
+		}
+		code := s.code(c)
+		if last == 0 {
+			b.WriteRune(c)
+			last = code
+			continue
+		}
+		if code == '0' || code == last {
+			continue
+		}
+		b.WriteRune(code)
+		last = code
+	}
+	if b.Len() == 0 {
+		return "", nil
+	}
+	for i := len([]rune(b.String())); i < 4; i++ {
```

Comment:
why 4 (because type rune = int32)?

Reply:
If a given string can be represented with soundex (if it has at lease one character listed in the `code` function), the resulting string can have an arbitrary length, but at least four characters:

```
mysql> select soundex('');
+-------------+
| soundex('') |
+-------------+
|             |
+-------------+
1 row in set (0,00 sec)

mysql> select soundex('Q');
+--------------+
| soundex('Q') |
+--------------+
| Q000         |
+--------------+
1 row in set (0,01 sec)

mysql> select soundex('Quadratically');
+--------------------------+
| soundex('Quadratically') |
+--------------------------+
| Q36324                   |
+--------------------------+
1 row in set (0,00 sec)
```

Reply:
(I am not sure if my previous comment got saved)

Even though the reference doesn't say so, MySQL returns at least four characters, if one of them is a letter:

```
mysql> select soundex('');
+-------------+
| soundex('') |
+-------------+
|             |
+-------------+
1 row in set (0,00 sec)

mysql> select soundex('Q');
+--------------+
| soundex('Q') |
+--------------+
| Q000         |
+--------------+
1 row in set (0,00 sec)

mysql> select soundex('Quadratically');
+--------------------------+
| soundex('Quadratically') |
+--------------------------+
| Q36324                   |
+--------------------------+
1 row in set (0,00 sec)
```

Reply:
I am not sure how close this should follow MySQL, since it is not a replacement. Always returning a string of 0 or 4 characters would be a possibility, too.